src/renderer/components/recording-page/recording-visualization.tsx

```diff
@@ -2,226 +2,161 @@ import log from 'electron-log';
-
-  // useEffect(() => {
-  //   const [waveformContainer, waveformImage, spectrogramContainer, spectrogramImage] = [
-  //     waveformContainerRef.current,
-  //     waveformImageRef.current,
-  //     spectrogramContainerRef.current,
-  //     spectrogramImageRef.current,
-  //   ];
-  //
-  //   if (waveformContainer && waveformImage && spectrogramContainer && spectrogramImage) {
-  //     const waveformCanvases = waveformContainer.getElementsByTagName('canvas');
-  //     if (waveformCanvases.length) {
-  //       log.silly('Updating waveform image');
-  //       const waveform: HTMLCanvasElement = waveformCanvases[waveformCanvases.length - 1];
-  //       waveformImage.src = waveform.toDataURL();
-  //     }
-  //     const spectrogramCanvases = spectrogramContainer.getElementsByTagName('canvas');
-  //     if (spectrogramCanvases.length) {
-  //       log.silly('Updating spectrogram image');
-  //       const spectrogram: HTMLCanvasElement = spectrogramCanvases[spectrogramCanvases.length - 1];
-  //       spectrogramImage.src = spectrogram.toDataURL();
-  //     }
-  //   }
-  // });
+  const [, , audioInputDeviceId, , ,] = useAudioInputOutputDevices();
+  const fileMonitor = new FileMonitor(basePath);
+
+  function checkSpectrogramPlugin() {
+    const waveSurfer = waveSurferRef.current;
+
+    if (!waveSurfer) {
+      return;
+    }
+
+    if (!('spectrogram' in waveSurfer.getActivePlugins())) {
+      waveSurfer
+        .addPlugin(
+          SpectrogramPlugin.create({
+            container: spectrogramContainerRef.current!,
+            labels: true,
+          }),
+        )
+        .initPlugin('spectrogram');
+    }
+  }
+
+  function loadWavFile(filePathToRead: string) {
+    const waveSurfer = waveSurferRef.current;
+
+    if (!waveSurfer) {
+      return noOp();
+    }
+    let cancelled = false;
+
+    (async (): Promise<void> => {
+      if ((await checkFileExistence(filePathToRead)) === 'file') {
+        const data = await readFile(filePathToRead, true);
+        if (!cancelled) {
+          waveSurfer.loadBlob(new Blob([new Uint8Array(data)]));
+          log.info('Triggered waveform update');
+        }
+      }
+    })();
+
+    return (): void => {
+      cancelled = true;
+    };
+  }
 
   useEffect(() => {
+    // create wavesurfer instance with plugins
     const waveSurfer = WaveSurfer.create({
       interact: false,
       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       container: waveformContainerRef.current!,
+      height: 100,
       plugins: [
         SpectrogramPlugin.create({
-          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
           container: spectrogramContainerRef.current!,
           labels: true,
         }),
+        MicrophonePlugin.create({}),
       ],
     });
+
+    // add watchers to the fileMonitor
+    fileMonitor.watch(['add', 'unlink', 'change']);
+    fileMonitor.getSubject().subscribe({
+      next: (x: string[]) => {
```

Comment:
Use tuple type to get more typing help from the compiler.

Reply:
Could explain/elaborate what you mean here?

Reply:
`x: [string, string]`

Reply:
are you refering to the fileMonitor.watch(['add', 'unlink', 'change']); line?

Reply:
`(x: string[]) => {` to `(x: [string, string]) => {`

Reply:
Also, you can just pass this function in. No need for the object.